panic: interface conversion: interface {} is nil, not string

Running into the below issue while running kubent, can someone help?

```
panic: interface conversion: interface {} is nil, not string

goroutine 1 [running]:
github.com/doitintl/kube-no-trouble/pkg/judge.(*RegoJudge).Eval(0xc00015fc10, 0xc000273500, 0x29c, 0x2a0, 0x29c, 0x2a0, 0x0, 0x1f4, 0xc0002f4390)
	/__w/kube-no-trouble/kube-no-trouble/pkg/judge/rego.go:76 +0x7c2
main.main()
	/__w/kube-no-trouble/kube-no-trouble/cmd/kubent/main.go:97 +0x3be
```

Go version - go1.15.5 darwin/amd64

kubent version - 0.3.2

Hi @stepanstipl thanks for the reply. I downloaded the tar file, unzipped it and ran the binary. Please find the full output below.

```
$ /usr/local/bin/kubent 
1:02PM INF >>> Kube No Trouble `kubent` <<<
1:02PM INF version 0.3.2 (git sha 919129b596890475965cda2b972cb6fded71f40b)
1:02PM INF Initializing collectors and retrieving data
1:02PM INF Retrieved 58 resources from collector name=Cluster
1:02PM INF Retrieved 994 resources from collector name="Helm v2"
1:02PM INF Retrieved 0 resources from collector name="Helm v3"
1:02PM INF Loaded ruleset name=deprecated-1-16.rego
1:02PM INF Loaded ruleset name=deprecated-1-22.rego
panic: interface conversion: interface {} is nil, not string

goroutine 1 [running]:
github.com/doitintl/kube-no-trouble/pkg/judge.(*RegoJudge).Eval(0xc000c54430, 0xc00005ea00, 0x41c, 0x4a0, 0x41c, 0x4a0, 0x0, 0x1f4, 0xc0002d84e0)
	/__w/kube-no-trouble/kube-no-trouble/pkg/judge/rego.go:76 +0x7c2
main.main()
	/__w/kube-no-trouble/kube-no-trouble/cmd/kubent/main.go:97 +0x3be
```

If I use the below command, I get the output without any error but I'm not sure if the list is exhaustive.

```
/usr/local/bin/kubent --helm2=false
```

Hi @stepanstipl thanks for the reply and apologies for the late reply, I missed the mail. I downloaded the tar file, unzipped it and ran the binary. Please find the full output below.

```
$ /usr/local/bin/kubent 
1:02PM INF >>> Kube No Trouble `kubent` <<<
1:02PM INF version 0.3.2 (git sha 919129b596890475965cda2b972cb6fded71f40b)
1:02PM INF Initializing collectors and retrieving data
1:02PM INF Retrieved 58 resources from collector name=Cluster
1:02PM INF Retrieved 994 resources from collector name="Helm v2"
1:02PM INF Retrieved 0 resources from collector name="Helm v3"
1:02PM INF Loaded ruleset name=deprecated-1-16.rego
1:02PM INF Loaded ruleset name=deprecated-1-22.rego
panic: interface conversion: interface {} is nil, not string

goroutine 1 [running]:
github.com/doitintl/kube-no-trouble/pkg/judge.(*RegoJudge).Eval(0xc000c54430, 0xc00005ea00, 0x41c, 0x4a0, 0x41c, 0x4a0, 0x0, 0x1f4, 0xc0002d84e0)
	/__w/kube-no-trouble/kube-no-trouble/pkg/judge/rego.go:76 +0x7c2
main.main()
	/__w/kube-no-trouble/kube-no-trouble/cmd/kubent/main.go:97 +0x3be
```

If I use the below command, I get the output without any error but I'm not sure if the list is exhaustive.

```
/usr/local/bin/kubent --helm2=false
```

I'm getting the same error using nightly-0.4.0-83-g055df4e

```
❌ . (none) kube-no-trouble ➤ ./kubent
5:28PM INF >>> Kube No Trouble `kubent` <<<
5:28PM INF version nightly-0.4.0-83-g055df4e (git sha 055df4e213a394ad107021724714c5130759b284)
5:28PM INF Initializing collectors and retrieving data
5:28PM INF Target K8s version is 1.15+
5:28PM WRN Failed to retrieve: networking.k8s.io/v1, Resource=ingressclasses: the server could not find the requested resource
5:28PM WRN Failed to retrieve: storage.k8s.io/v1, Resource=csidrivers: the server could not find the requested resource
5:28PM WRN Failed to retrieve: storage.k8s.io/v1, Resource=csinodes: the server could not find the requested resource
5:28PM WRN Failed to retrieve: authorization.k8s.io/v1, Resource=subjectaccessreviews: the server does not allow this method on the requested resource
5:28PM WRN Failed to retrieve: authorization.k8s.io/v1, Resource=selfsubjectaccessreviews: the server does not allow this method on the requested resource
5:28PM WRN Failed to retrieve: authorization.k8s.io/v1, Resource=localsubjectaccessreviews: the server could not find the requested resource
5:28PM WRN Failed to retrieve: authentication.k8s.io/v1, Resource=tokenreviews: the server does not allow this method on the requested resource
5:28PM WRN Failed to retrieve: certificates.k8s.io/v1, Resource=certificatesigningrequests: the server could not find the requested resource
5:28PM WRN Failed to retrieve: apiextensions.k8s.io/v1, Resource=customresourcedefinitions: the server could not find the requested resource
5:28PM WRN Failed to retrieve: admissionregistration.k8s.io/v1, Resource=mutatingwebhookconfigurations: the server could not find the requested resource
5:28PM WRN Failed to retrieve: admissionregistration.k8s.io/v1, Resource=validatingwebhookconfigurations: the server could not find the requested resource
5:28PM WRN Failed to retrieve: node.k8s.io/v1, Resource=runtimeclasses: the server could not find the requested resource
5:28PM WRN Failed to retrieve: policy/v1, Resource=poddisruptionbudgets: the server could not find the requested resource
5:28PM WRN Failed to retrieve: discovery.k8s.io/v1, Resource=endpointslices: the server could not find the requested resource
5:28PM WRN Failed to retrieve: batch/v1, Resource=cronjobs: the server could not find the requested resource
5:28PM INF Retrieved 58 resources from collector name=Cluster
5:28PM INF Retrieved 939 resources from collector name="Helm v2"
5:28PM INF Retrieved 24 resources from collector name="Helm v2"
5:28PM INF Loaded ruleset name=custom.rego.tmpl
5:28PM INF Loaded ruleset name=deprecated-1-16.rego
5:28PM INF Loaded ruleset name=deprecated-1-22.rego
5:28PM INF Loaded ruleset name=deprecated-1-25.rego
panic: interface conversion: interface {} is nil, not string

goroutine 1 [running]:
github.com/doitintl/kube-no-trouble/pkg/judge.(*RegoJudge).Eval(0xc00063cc40, 0xc0011e0000, 0x3fd, 0x400, 0xc00063cc40, 0x0, 0x0, 0x0, 0xc0001c3350)
        /__w/kube-no-trouble/kube-no-trouble/pkg/judge/rego.go:55 +0x94d
main.main()
        /__w/kube-no-trouble/kube-no-trouble/cmd/kubent/main.go:135 +0x636
```

``` bash
4:42PM TRC parsing +map[ApiVersion:extensions/v1beta1 Kind:DaemonSet Name:cattle-node-agent Namespace:cattle-system ReplaceWith:apps/v1 RuleSet:Deprecated APIs removed in 1.16 Since:1.9]
4:42PM TRC parsing +map[ApiVersion:extensions/v1beta1 Kind:Deployment Name:cattle-cluster-agent Namespace:cattle-system ReplaceWith:apps/v1 RuleSet:Deprecated APIs removed in 1.16 Since:1.9]
4:42PM TRC parsing +map[ApiVersion:apps/v1beta2 Kind:StatefulSet Name:azure-agent Namespace:azure-agent ReplaceWith:apps/v1 RuleSet:Deprecated APIs removed in 1.16 Since:1.9]
4:42PM TRC parsing +map[ApiVersion:extensions/v1beta1 Kind:DaemonSet Name:collection-fluent-bit Namespace:sumologic ReplaceWith:apps/v1 RuleSet:Deprecated APIs removed in 1.16 Since:1.9]
4:42PM TRC parsing +map[ApiVersion:extensions/v1beta1 Kind:DaemonSet Name:collection-falco Namespace:sumologic ReplaceWith:apps/v1 RuleSet:Deprecated APIs removed in 1.16 Since:1.9]
4:42PM TRC parsing +map[ApiVersion:apps/v1beta1 Kind:StatefulSet Name:rcm-journal-reporting-hardclose-service Namespace:<nil> ReplaceWith:apps/v1 RuleSet:Deprecated APIs removed in 1.16 Since:1.9]
panic: interface conversion: interface {} is nil, not string
```

--- a/pkg/collector/helm.go
+++ b/pkg/collector/helm.go
@@ -33,7 +33,7 @@ func fixNamespace(manifest *map[string]interface{}, defaultNamespace string) {
 	if meta, ok := (*manifest)["metadata"]; ok {
 		switch v := meta.(type) {
 		case map[string]interface{}:
-			if _, ok := v["namespace"]; !ok {
+			if val, ok := v["namespace"]; !ok || val == nil {
 				v["namespace"] = defaultNamespace
 			}
 		}

--- a/pkg/collector/helm.go
+++ b/pkg/collector/helm.go
@@ -1,0 +1,41 @@
+package collector
+
+import (
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"helm.sh/helm/v3/pkg/releaseutil"
+)
+
+func parseManifests(manifest string, defaultNamespace string) ([]map[string]interface{}, error) {
+	var results []map[string]interface{}
+
+	manifests := releaseutil.SplitManifests(manifest)
+	for i, m := range manifests {
+		var manifest map[string]interface{}
+
+		err := yaml.Unmarshal([]byte(m), &manifest)
+		if err != nil {
+			err = fmt.Errorf("failed to parse manifest %s: %v", i, err)
+			return nil, err
+		}
+
+		fixNamespace(&manifest, defaultNamespace)
+
+		results = append(results, manifest)
+	}
+
+	return results, nil
+}
+
+func fixNamespace(manifest *map[string]interface{}, defaultNamespace string) {
+	// Default to the release namespace if the manifest doesn't have the namespace set
+	if meta, ok := (*manifest)["metadata"]; ok {
+		switch v := meta.(type) {
+		case map[string]interface{}:
+			if _, ok := v["namespace"]; !ok {
+				v["namespace"] = defaultNamespace
+			}
+		}
+	}
+}

--- a/pkg/collector/helm2.go
+++ b/pkg/collector/helm2.go
@@ -1,11 +1,9 @@
 package collector
 
 import (
-	"github.com/ghodss/yaml"
 	"github.com/rs/zerolog/log"
 	"helm.sh/helm/pkg/storage"
 	"helm.sh/helm/pkg/storage/driver"
-	"helm.sh/helm/v3/pkg/releaseutil"
 	"k8s.io/client-go/discovery"
 	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
 )
@@ -70,27 +68,10 @@ func (c *HelmV2Collector) Get() ([]map[string]interface{}, error) {
 	var results []map[string]interface{}
 
 	for _, r := range releases {
-		manifests := releaseutil.SplitManifests(r.Manifest)
-		for _, m := range manifests {
-			var manifest map[string]interface{}
-
-			err := yaml.Unmarshal([]byte(m), &manifest)
-			if err != nil {
-				log.Warn().Msgf("failed to parse release %s/%s: %v", r.Namespace, r.Name, err)
-				continue
-			}
-
-			// Default to the release namespace if the manifest doesn't have the namespace set
-			if meta, ok := manifest["metadata"]; ok {
-				switch v := meta.(type) {
-				case map[string]interface{}:
-					if _, ok := v["namespace"]; !ok {
-						v["namespace"] = r.Namespace
-					}
-				}
-			}
-
-			results = append(results, manifest)
+		if manifests, err := parseManifests(r.Manifest, r.Namespace); err != nil {
+			log.Warn().Msgf("failed to parse release %s/%s: %v", r.Namespace, r.Name, err)
+		} else {
+			results = append(results, manifests...)
 		}
 	}
 

--- a/pkg/collector/helm3.go
+++ b/pkg/collector/helm3.go
@@ -1,9 +1,7 @@
 package collector
 
 import (
-	"github.com/ghodss/yaml"
 	"github.com/rs/zerolog/log"
-	"helm.sh/helm/v3/pkg/releaseutil"
 	"helm.sh/helm/v3/pkg/storage"
 	"helm.sh/helm/v3/pkg/storage/driver"
 	"k8s.io/client-go/discovery"
@@ -69,27 +67,10 @@ func (c *HelmV3Collector) Get() ([]map[string]interface{}, error) {
 	var results []map[string]interface{}
 
 	for _, r := range releases {
-		manifests := releaseutil.SplitManifests(r.Manifest)
-		for _, m := range manifests {
-			var manifest map[string]interface{}
-
-			err := yaml.Unmarshal([]byte(m), &manifest)
-			if err != nil {
-				log.Warn().Msgf("failed to parse release %s/%s: %v", r.Namespace, r.Name, err)
-				continue
-			}
-
-			// Default to the release namespace if the manifest doesn't have the namespace set
-			if meta, ok := manifest["metadata"]; ok {
-				switch v := meta.(type) {
-				case map[string]interface{}:
-					if _, ok := v["namespace"]; !ok {
-						v["namespace"] = r.Namespace
-					}
-				}
-			}
-
-			results = append(results, manifest)
+		if manifests, err := parseManifests(r.Manifest, r.Namespace); err != nil {
+			log.Warn().Msgf("failed to parse release %s/%s: %v", r.Namespace, r.Name, err)
+		} else {
+			results = append(results, manifests...)
 		}
 	}
 

--- a/pkg/collector/helm_test.go
+++ b/pkg/collector/helm_test.go
@@ -1,0 +1,87 @@
+package collector
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestSplitManifests(t *testing.T) {
+	testCases := []struct {
+		name     string
+		input    string // manifest
+		expected int    // kinds of objects
+	}{
+		{"single",
+			"abc: x",
+			1,
+		},
+		{"multiple",
+			"abc: x\n---\ndef: y",
+			2,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+
+			manifests, err := parseManifests(tc.input, "default")
+
+			if err != nil {
+				t.Fatalf("Expected to successfully parse manifests: %v", err)
+			}
+
+			if len(manifests) != tc.expected {
+				t.Fatalf("Expected %d manifests, instead got: %d", tc.expected, len(manifests))
+			}
+
+		})
+	}
+}
+
+func TestFixNamespace(t *testing.T) {
+	testCases := []struct {
+		name     string
+		input    map[string]interface{} // manifest
+		expected string                 // kinds of objects
+	}{
+		{"present",
+			map[string]interface{}{"metadata": map[string]interface{}{"namespace": "some-namespace"}},
+			"some-namespace",
+		},
+		{"missing",
+			map[string]interface{}{"metadata": map[string]interface{}{}},
+			"default-namespace",
+		},
+		{"nil",
+			map[string]interface{}{"metadata": map[string]interface{}{"namespace": nil}},
+			"default-namespace",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+
+			fixNamespace(&tc.input, tc.expected)
+
+			meta, ok := tc.input["metadata"]
+			if !ok {
+				t.Fatalf("Expected fixed manifest to have metadata key")
+			}
+
+			expectedType := "map[string]interface {}"
+			if reflect.TypeOf(meta).String() != expectedType {
+				t.Fatalf("Expected metadata type to be %s, instead got: %T", expectedType, meta)
+			}
+
+			actual, ok := meta.(map[string]interface{})["namespace"]
+			if !ok {
+				t.Fatalf("Expected fixed manifest to have metadata.namespace key")
+			}
+
+			if actual != tc.expected {
+				t.Fatalf("Expected namespace to be: %s, instead got: %s", tc.expected, actual)
+			}
+
+		})
+	}
+}

